chiron/multistate.py

```diff
@@ -0,0 +1,968 @@
+        )
+
+    def _update_timing(
+        self, iteration_time, partial_total_time, run_initial_iteration, iteration_limit
+    ):
+        """
+        Function that computes and transmits timing information to reporter.
+
+        Parameters
+        ----------
+        iteration_time : float
+            Time took in the iteration.
+        partial_total_time : float
+            Partial total time elapsed.
+        run_initial_iteration : int
+            Iteration where to start/resume the simulation.
+        iteration_limit : int
+            Hard limit on number of iterations to be run by the sampler.
+        """
+        self._timing_data["iteration_seconds"] = iteration_time
+        self._timing_data["average_seconds_per_iteration"] = partial_total_time / (
+            self._iteration - run_initial_iteration
+        )
+        estimated_timedelta_remaining = datetime.timedelta(
+            seconds=self._timing_data["average_seconds_per_iteration"]
+            * (iteration_limit - self._iteration)
+        )
+        estimated_finish_date = datetime.datetime.now() + estimated_timedelta_remaining
+        self._timing_data["estimated_time_remaining"] = str(
+            estimated_timedelta_remaining
+        )  # Putting it in dict as str
+        self._timing_data[
+            "estimated_localtime_finish_date"
+        ] = estimated_finish_date.strftime("%Y-%b-%d-%H:%M:%S")
+        total_time_in_seconds = datetime.timedelta(
+            seconds=self._timing_data["average_seconds_per_iteration"] * iteration_limit
+        )
+        self._timing_data["estimated_total_time"] = str(total_time_in_seconds)
+
+        # Estimate performance
+        moves_iterator = self._flatten_moves_iterator()
+        # Only consider "dynamic" moves (timestep and n_steps attributes)
+        moves_times = [
+            move.timestep.value_in_unit(unit.nanosecond) * move.n_steps
+            for move in moves_iterator
+            if hasattr(move, "timestep") and hasattr(move, "n_steps")
+        ]
+        iteration_simulated_nanoseconds = sum(moves_times)
+        seconds_in_a_day = (1 * unit.day).value_in_unit(unit.seconds)
+        self._timing_data["ns_per_day"] = iteration_simulated_nanoseconds / (
+            self._timing_data["average_seconds_per_iteration"] / seconds_in_a_day
+        )
+
+    def _flatten_moves_iterator(self):
+        """Recursively flatten MCMC moves. Handles the cases where each move can be a set of moves, for example with
+        SequenceMove or WeightedMove objects."""
+
+        def flatten(iterator):
+            try:
+                yield from [
+                    inner_move for move in iterator for inner_move in flatten(move)
+                ]
+            except TypeError:  # Inner object is not iterable, finish flattening.
+                yield iterator
+
+        return flatten(self.mcmc_moves)
+
+    def _check_nan_energy(self):
```

Comment:
Is this still useful for us? We may want to get rid of it and introduce our own checks for stability, which may be different. 

We can also just check for NaNs during the energy matrix computation.

Whatever we do, we will want to make sure this is still efficient and can be `jit`ted.

Reply:
This is removed now.